### ISSUE TITLE
DensifyPointCloud: Speed up tips & tricks

### ISSUE THREAD

**Issue by amughrabi** [cdcseacave/openMVS#1048]
Thanks a ton for the beautiful work! 

I want to ask if we have any tips or tricks to speed up `DensifyPointCloud` with masking. I have a scene containing 268 images with 4624x3468 resolutions for each image. Resizing the images may lead to missing segments in the masks. FWIW, most processing time goes to generating and filtering depth maps (~45mins). 

```bash
DensifyPointCloud scene.mvs --mask-path masks_omvs/ --ignore-mask-label 255 --filter-point-cloud 1 
```

**Comment by cdcseacave** [comment]
you can use `--max-resolution 1024` or smaller

**Comment by amughrabi** [comment]
Thanks for the quick answer! 

It is dramatically boosted to 6m32s!

Upon examining the PointCloud construction, I noticed that there might be a correlation between the holes and the `--max-resolution` parameter. Are there any other parameters that I want to consider when I change  `--max-resolution`?

`--max-resolution 512`
![512](https://github.com/cdcseacave/openMVS/assets/8401456/4dbf54f5-66db-4ba5-b811-c6c7272f8856)

`--max-resolution 1024`

![1024](https://github.com/cdcseacave/openMVS/assets/8401456/4303963e-06d5-4439-a4cd-7b5861ef0cb5)


`--max-resolution 2560` (default)
![default-2560](https://github.com/cdcseacave/openMVS/assets/8401456/d5c29445-65fc-4944-86b4-f1cb205b030c)

**Comment by cdcseacave** [comment]
this is expected for texturesless surfaces

**Comment by amughrabi** [comment]
@cdcseacave -  are there other parameters/flags I can consider for further speed optimisations?

**Comment by 4CJ7T** [comment]
No, but if you were to use scene clusters and recompute depth maps for each scene because of limited storage space, multiple NVIDIA GPUs could be scripted.

**Comment by amughrabi** [comment]
@4CJ7T - thanks for your reply! AFAICT, I am not using the scene clusters, but I am still writing 286 depth maps with filtered ones to the disk. Later, we removed them using `-remove-dmaps`, which sounds inefficient. Is there a way to keep them in memory instead?

**Comment by 4CJ7T** [comment]
All depth maps are loaded into memory for fusion. `--remove-dmaps` removes depth maps after fusion to clear storage space.

**Comment by amughrabi** [comment]
Sounds reasonable. Can you advise how to stop writing depth maps to disk?

**Comment by 4CJ7T** [comment]
Currently there's no option for that. If depth maps were stored in memory geometric consistency would double memory usage but as long as scenes are small enough it might be alright.

**Comment by amughrabi** [comment]
> There is no option for that currently.

Can you please point me out where can I do this modification?

**Comment by cdcseacave** [comment]
writing DMAPs to disk is usually necessary, unless you have infinite memory

**Comment by amughrabi** [comment]
> writing DMAPs to disk is usually necessary, unless you have infinite memory

AFAICT, It depends on the use case. For the small scenes (~50 depth maps), it is useful not to write them on the disk for time-critical applications. I want to avoid the IO. The memory is not a problem. Can we at least have a flag to stop writing?

**Comment by cdcseacave** [comment]
it is not so easy, you need to change the code to add this option, each time it tries to save / load from disk

**Comment by amughrabi** [comment]
> it is not so easy, you need to change the code to add this option, each time it tries to save / load from disk

Very well, @cdcseacave! I can put a hand on this, but I want some guidance.

**Comment by cdcseacave** [comment]
The idea is simple, the depth-map data needs to be available during the various stages, so if you do not save it to disk, you need not to release it either. Find all places where `DepthData` is serialized to DMAP file, and disable both saving and releasing it.

**Comment by cdcseacave** [comment]
Note that with the recent dense reconstruction overhaul (PR #1246), depth-map caching and pipelining have been substantially enhanced:
- Confidence adjustment now recalibrates directly in resident memory buffers, eliminating the intermediate disk round-trips (\.adjusted.cmap\ / \.adjusted.fast.cmap\).
- Depth-map caching (\DMapCache\) caches loaded depth maps in memory across references to reduce disk reads during fusion.

Closing this discussion as answered.
